specification/sections/02-providers.md

```diff
@@ -26,6 +26,22 @@ Providers are the "translator" between the flag evaluation calls made in applica
 provider.getMetadata().getName(); // "my-custom-provider"
 ```
 
+#### Requirement 2.1.2
+
+> The provider `metadata` member or accessor **MAY** define an `allFlagKeys` field or accessor of type collection of strings, which identifies all available flag keys in the provider.
+
+```typescript
+provider.getMetadata().getAllFlagKeys(); // ['featureA', 'featureB']
```

Comment:
I may be misunderstanding the use-case here, but I believe these also need typing information, else you won't know which method to call to get the value out, right?

Reply:
I am also having a bit of a hard time understanding the ergonomics of this. This does satisfy one category of usage, "does the flag I plan on evaluating exist", but it doesn't provide a great means of getting all of the values associated with those keys.

There are a couple use cases I am aware of for evaluating all the flags.

1. Bootstrap. You evaluate all the flags so you can use them for something like SSR. You render the values into the page and you avoid pop-in. Or you store them into a cache when you build your app for the same reason.

2. You evaluate them all and then you pluck values out of the map, versus evaluating them individually. One useful example of this is to ensure that all your flags are consistent with each other for some transactional period, like handling an http request. (The flags could change part way through the request, and if the flags have relationships, then that isn't ideal.)

For me the second one is somewhat of an anti-pattern, or shows an underlying weakness of a system. Once you do that, then you lose insight into what flags are actually being evaluated, which causes problems for tracking exposure for things like an experiment.

Reply:
I am not sure how we get from the keys, to either of these use-cases, easily.

Reply:
Those are great points, especially for how we'd go from keys to the correct resolution method. I'm hoping to avoid adding a complete bulk evaluation right now. Based on the discussion predating this, we want to try a more minimal approach at first.

We could be more explicit about some required fields for flag metadata, including key and type. Would that be sufficient information if we return a collection of flag metadata through the provider metadata?

We still wouldn't meet the bootstrapping use case described by @kinyoklion, but we would be able to

1. Detect stale flags being evaluated and
2. Build a debug screen based on available flags

which are two use cases we've encountered.

Reply:
I think it would be worth adding additional non-normative text explaining the purpose of this feature and some example supported use-cases.

Reply:
I took an initial stab at including more text around this feature. Let me know of any edits/clarifications I can make there.